vpp/oclpostprocess_transform.cpp

```diff
@@ -0,0 +1,281 @@
+ *    Author: Jia Meng<jia.meng@intel.com>
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Lesser General Public License
+ *  as published by the Free Software Foundation; either version 2.1
+ *  of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with this library; if not, write to the Free
+ *  Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ *  Boston, MA 02110-1301 USA
+ */
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "oclpostprocess_transform.h"
+#include "vaapipostprocess_factory.h"
+#include "common/common_def.h"
+#include "common/log.h"
+#include "ocl/oclcontext.h"
+#include "vaapi/vaapiutils.h"
+#include "vpp/oclvppimage.h"
+
+namespace YamiMediaCodec {
+
+YamiStatus
+OclPostProcessTransform::process(const SharedPtr<VideoFrame>& src,
+    const SharedPtr<VideoFrame>& dst)
+{
+    YamiStatus status = ensureContext("transform");
+    if (status != YAMI_SUCCESS)
+        return status;
+
+    if (src->fourcc != YAMI_FOURCC_NV12 || dst->fourcc != YAMI_FOURCC_NV12) {
+        ERROR("only support transform of NV12 video frame");
+        return YAMI_INVALID_PARAM;
+    }
+
+    cl_image_format format;
+    format.image_channel_order = CL_RGBA;
+    format.image_channel_data_type = CL_UNORM_INT8;
+    SharedPtr<OclVppCLImage> srcImage = OclVppCLImage::create(m_display, src, m_context, format);
+    if (!srcImage) {
+        ERROR("failed to create cl image from src video frame");
+        return YAMI_FAIL;
+    }
+    SharedPtr<OclVppCLImage> dstImage = OclVppCLImage::create(m_display, dst, m_context, format);
+    if (!dstImage) {
+        ERROR("failed to create cl image from dst video frame");
+        return YAMI_FAIL;
+    }
+
+    if (m_transform & VPP_TRANSFORM_FLIP_H && m_transform & VPP_TRANSFORM_FLIP_V) {
+        // flip both v and h is effectively rotate 180
+        m_transform &= ~(VPP_TRANSFORM_FLIP_H | VPP_TRANSFORM_FLIP_V);
+        switch (m_transform) {
+        case VPP_TRANSFORM_ROT_90:
+            m_transform = VPP_TRANSFORM_ROT_270;
+            break;
+        case VPP_TRANSFORM_ROT_180:
+            m_transform = 0;
+            break;
+        case VPP_TRANSFORM_ROT_270:
+            m_transform = VPP_TRANSFORM_ROT_90;
+            break;
+        default:
+            m_transform = VPP_TRANSFORM_ROT_180;
+            break;
+        }
+    }
+
+    bool bFlip = m_transform & (VPP_TRANSFORM_FLIP_H | VPP_TRANSFORM_FLIP_V);
+    bool bRotate = m_transform & (VPP_TRANSFORM_ROT_90 | VPP_TRANSFORM_ROT_180 | VPP_TRANSFORM_ROT_270);
+
+    if (bFlip && bRotate) {
```

Comment:
I intended to make the api flexible, so user cat toggle single or multiple flags in setParameters(). After your suggestion, I realize that the combined case can be optimized by adding 2 more ocl kernels, and also scratch buffer is unnecessary.

There are totally 7 cases.
case 1 (new kernel): FH+R90, FV+R270, R90+FV, R270+FH
case 2 (new kernel): FV+R90, FH+R270, R90+FH, R270+FV
case 3: FV, FH+R180, R180+FH
case 4: FH, FV+R180, R180+FV
case 5: R90
case 6: R180, FH+FV, FV+FH 
case 7: R270

I also assume Flip is performed before Rotate if user set both flags. So need to add a new private function flipAndRotate().